File: 289-game-of-life/game-of-life.py
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        
        def countNeighbours(r,c):
            nei=0
            for i in range(r-1,r+2):
                for j in range(c-1,c+2):
                    if((i==r and j==c) or i<0 or j<0 or i==rows or j==cols):
                        continue
                    if(board[i][j] in [1,3]):
                        nei+=1
            return nei;

        rows,cols=len(board),len(board[0])
        for r in range(rows):
            for c in range(cols):
                nei=countNeighbours(r,c)
                if board[r][c]==1:
                    if nei in [2,3]:
                        board[r][c]=3
                elif nei==3:
                        board[r][c]=2

        for r in range(rows):
            for c in range(cols):
                if(board[r][c]==1):
                    board[r][c]=0
                elif board[r][c] in [2,3]:
                    board[r][c]=1

File: 289-game-of-life/game-of-life.java
class Solution {
    public void gameOfLife(int[][] board) {
        int r=board.length;
        int c=board[0].length;
        int[][] copyboard=new int[r][c];
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                copyboard[i][j]=board[i][j];
            }
        }
        int[] dx={1,1,0,-1,-1,-1,0,1};
        int[] dy={0,1,1,1,0,-1,-1,-1};
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                int livecounts=0;
                for(int k=0;k<8;k++){
                    int nrow=i+dx[k];
                    int ncol=j+dy[k];
                    if(isSafe(nrow,ncol,r,c) && copyboard[nrow][ncol]==1){
                        livecounts++;
                    }
                }
                if(copyboard[i][j]==0 && livecounts==3){
                    board[i][j]=1;
                }if(copyboard[i][j]==1 && livecounts<2 || livecounts>3){
                    board[i][j]=0;
                }
            }
        }
    }
    private boolean isSafe(int x,int y,int r,int c){
        return (x>=0 && x<r && y>=0 && y<c);
    }
}
